docs/spec/constructors.rst

```diff
@@ -0,0 +1,487 @@
+does not inherit either of these methods from a base class other than
+``object``, a type checker should evaluate the argument list using the
+``__new__`` and ``__init__`` methods from the ``object`` class.
+
+  ::
+
+    class MyClass5:
+        pass
+    
+    MyClass5()  # OK
+    MyClass5(1)  # Type error
+
+
+Constructor Calls for type[T]
+-----------------------------
+
+When a value of type ``type[T]`` (where ``T`` is a type variable or a concrete
+class) is called, a type checker should evaluate the constructor call as if
+it is being made on the class ``T`` (or the class that represents the upper bound
+of type variable ``T``). This means the type checker should use the ``__call__``
+method of ``T``'s metaclass and the ``__new__`` and ``__init__`` methods of ``T``
+to evaluate the constructor call.
+
+It should be noted that such code could be unsafe because the type ``type[T]``
+may represent subclasses of ``T``, and those subclasses could redefine the
+``__new__`` and ``__init__`` methods in a way that is incompatible with the
+base class. Likewise, the metaclass of ``T`` could redefine the ``__call__``
+method in a way that is incompatible with the base metaclass.
+
+
+Specialization During Construction
+----------------------------------
+
+As discussed above, if a class is generic and not explicitly specialized, its
+type variables should be solved using the arguments passed to the ``__new__``
+and ``__init__`` methods. If one or more type variables are not solved during
+these method evaluations, they should take on their default values.
+
+  ::
+
+    T1 = TypeVar("T1")
+    T2 = TypeVar("T2")
+    T3 = TypeVar("T3", default=str)
+
+    class MyClass1(Generic[T1, T2]):
+        def __new__(cls, x: T1) -> Self: ...
+
+    assert_type(MyClass1(1), MyClass1[int, Any])
+
+    class MyClass2(Generic[T1, T3]):
+        def __new__(cls, x: T1) -> Self: ...
+
+    assert_type(MyClass2(1), MyClass2[int, str])
+
+
+Consistency of ``__new__`` and ``__init__``
+-------------------------------------------
+
+Type checkers may optionally validate that the ``__new__`` and ``__init__``
+methods for a class have consistent signatures.
+
+  ::
+
+    class MyClass:
+        def __new__(cls) -> Self:
+            return super().__new__(cls)
+
+        # Type error: __new__ and __init__ have inconsistent signatures
+        def __init__(self, x: str) -> None:
+            pass
+
+
+Converting a Constructor to Callable
+------------------------------------
+
+Class objects are callable, which means they are compatible with callable types.
+
+  ::
+
+    def accepts_callable[**P, R](cb: Callable[P, R]) -> Callable[P, R]:
+        return cb
+
+    class MyClass:
+        def __init__(self, x: int) -> None:
+            pass
+
+    reveal_type(accepts_callable(MyClass))  # ``def (x: int) -> MyClass``
+
+When converting a class to a callable type, a type checker should use the
+following rules:
```

Comment:
The steps below looks awfully similar to the process for type-checking a constructor call. Are there subtle differences? Would there be a way to unify the two descriptions, making the spec less wordy, less repetitive, and easier to understand (no worries about subtle differences)?

Reply:
I've updated the introductory sentence to this section to indicate that these rules reflect the same rules specified above for evaluating constructor calls.